lib/ProductOpener/Ingredients.pm

```diff
@@ -3661,6 +3669,413 @@ sub replace_additive($$$) {
+			"réglisse",
+			"romarin",
+			"rose",
+			"rhum",
+			"sauge",
+			"saumon",
+			"sureau",
+			"thé",
+			"thym",
+			"vanille",
+			"vanille de Madagascar",
+			"autres agrumes",
+		]
+	],
+
+	[
+		[
+			"carbonate",
+			"carbonates acides",
+			"chlorure",
+			"citrate",
+			"iodure",
+			"nitrate",
+			"diphosphate",
+			"diphosphate",
+			"phosphate",
+			"sélénite",
+			"sulfate",
+			"hydroxyde",
+			"sulphate",
+		],
+		[
+			"aluminium",
+			"ammonium",
+			"calcium",
+			"cuivre",
+			"fer",
+			"magnésium",
+			"manganèse",
+			"potassium",
+			"sodium",
+			"zinc",
+		]
+	],
+],
+
+ru =>
+[
+	# oils
+	[
+		# categories
+		[
+			"масло",
+			"масло растительное",
+		],
+		# types
+		[
+			"подсолнечное",
+			"соевое",
+		],
+	],
+],
+
+);
+
+# Symbols to indicate labels like organic, fairtrade etc.
+my @symbols = ('\*\*\*', '\*\*', '\*', '°°°', '°°', '°', '\(1\)', '\(2\)', '¹', '²');
+my $symbols_regexp = join('|', @symbols);
+
+sub develop_ingredients_categories_and_types ($$) {
+	
+	my $product_lc = shift;
+	my $text = shift;	
+	
+	if (defined $ingredients_categories_and_types{$product_lc}) {
+
+		foreach my $categories_and_types_ref (@{$ingredients_categories_and_types{$product_lc}}) {
+
+			my $category_regexp = "";
+			foreach my $category (@{$categories_and_types_ref->[0]}) {
+				$category_regexp .= '|' . $category . '|' . $category . 's';
+				my $unaccented_category = unac_string_perl($category);
+				if ($unaccented_category ne $category) {
+					$category_regexp .= '|' . $unaccented_category . '|' . $unaccented_category . 's';
+				}
+
+			}
+			$category_regexp =~ s/^\|//;
+
+			if ($product_lc eq "en") {
+				$category_regexp = '(?:organic |fair trade )*(?:' . $category_regexp . ')(?:' . $symbols_regexp . ')*';
+			}
+			elsif ($product_lc eq "fr") {
+				$category_regexp = '(?:' . $category_regexp . ')(?: bio| biologique| équitable|s|\s|' . $symbols_regexp . ')*';
+			}
+			else {
+				$category_regexp = '(?:' . $category_regexp . ')(?:' . $symbols_regexp . ')*';
+			}
+
+			my $type_regexp = "";
+			foreach my $type (@{$categories_and_types_ref->[1]}) {
+				$type_regexp .= '|' . $type . '|' . $type . 's';
+				my $unaccented_type = unac_string_perl($type);
+				if ($unaccented_type ne $type) {
+					$type_regexp .= '|' . $unaccented_type . '|' . $unaccented_type . 's';
+				}
+
+			}
+			$type_regexp =~ s/^\|//;
+
+			# arôme naturel de citron-citron vert et d'autres agrumes
+			# -> separate types
+			$text =~ s/($type_regexp)-($type_regexp)/$1, $2/g;
+			
+			my $and = ' - ';
+			if (defined $and{$product_lc}) {
+				$and = $and{$product_lc};
+			}			
+			my $of = ' - ';
+			if (defined $of{$product_lc}) {
+				$of = $of{$product_lc};
+			}
+			my $and_of = ' - ';
+			if (defined $and_of{$product_lc}) {
+				$and_of = $and_of{$product_lc};
+			}
+			my $and_or = ' - ';
+			if (defined $and_or{$product_lc}) {
+				$and_or = $and_or{$product_lc};
+			}	
+			
+			if (($product_lc eq "en") or ($product_lc eq "ru")) {
```

Comment:
Do you mind if logic for English and Russian would be a bit different?
In Russian we never say "подсолнечное масло растительное" ("sunflower oil made of vegetables"), instead we say "масло растительное подсолнечное" ("vegetable oil made of sunflowers").
So instead of converting `масло растительное (подсолнечное)` into `подсолнечное масло растительное`, I think it should be converted into `масло растительное подсолнечное`.

On the other hand, although current expanded phrasing sounds a bit odd for a Russian ear, it definitely would be properly understood and current patch solves the problem of #5017 
So it'd be ok if the patch was merged like that :)

Reply:
We definitely should try to convert to what sounds best in Russian.

I was looking at https://ru.openfoodfacts.org/ingredients?filter=%D0%BC%D0%B0%D1%81%D0%BB%D0%BE and my impression was that in most cases масло was the last word. e.g. Пальмовое масло

In English we convert "Oil (Palm, Olive)" in "Palm oil, olive oil".
In French we do the reverse: "Huile (palme, olive)" becomes "Huile de palme, huile d'olive" (litterrally "oil of palm, oil of olive").
In Russian, what should we convert "Масло (пальмовое, оливковое)" to?
And what about "масло растительное (пальмовое, оливковое)" ?

I obviously don't speak Russian at all, so please tell me what's the best way to transform it. We can certainly improve the logic.

Reply:
> I was looking at https://ru.openfoodfacts.org/ingredients?filter=%D0%BC%D0%B0%D1%81%D0%BB%D0%BE and my impression was that in most cases масло was the last word. e.g. Пальмовое масло

You're absolutely correct about most common cases.

In Russian if the phrase has only word `масло` (`oil`), then the order does not matter (`подсолнечное масло` has absolutely same meaning as `масло подсолнечное`). That's the case you've described.

But if the phrase has 2 words instead of 1: `масло растительное` (`oil made from vegetables`) or `растительное масло` (`vegetable oil`), then the location of `подсолнечное` (`sunflower`) starts to matter a bit:
1. `oil` first: `масло растительное подсолнечное` and `растительное масло подсолнечное` both sound natural to a Russian ear (literal meaning: `oil made of vegetable - sunflower` and `vegetable oil made of sunflowers`).
2. `sunflower` first: `подсолнечное растительное масло` (`sunflower vegetable oil`) is also okay, but `подсолнечное масло растительное` sounds weird (literal meaning: `sunflower oil made of vegetables`).

So when `oil`/`vegetable oil` is put first, all phrases sound okay. But when `sunflower` (or any other type) is put first, then there's a case which sounds weird.

> In Russian, what should we convert "Масло (пальмовое, оливковое)" to?
> And what about "масло растительное (пальмовое, оливковое)" ?

The most simple solution would be to always put `oil`/`vegetable oil` before oil type:
- `Масло (пальмовое, оливковое)` -> `Масло пальмовое`, `масло оливковое`
- `масло растительное (пальмовое, оливковое)` -> `масло растительное пальмовое`, `масло растительное оливковое`

Reply:
@blazern Thanks a lot for the explanation. I made changes so that we always put the category first (oil) and then the type (palm).

How does it look now?

It implies that we need to make sure we have synonyms in the ingredients taxonomy so that entries like масло растительное пальмовое, масло растительное оливковое are recognized.

Reply:
> How does it look now?

Looks good! Thank you!

> It implies that we need to make sure we have synonyms in the ingredients taxonomy so that entries like масло растительное пальмовое, масло растительное оливковое are recognized.

I'm not sure I get it.
Do you mean that Russian oils should move from the `unknown` section of ingredients list into `ingredients.txt` file? :)